ZarrStorageArray and refactorings

normanrz opened the pull request:
### Description:
- Adds the `StorageArray` abstraction for WKW and Zarr
- Changes `block_len` to `chunk_size` and `file_len` to `chunks_per_shard`, but with backwards-compatible support
- Uses `pathlib` in more locations instead of string paths. This can be seen as a preparation for using [`upath`](https://github.com/fsspec/universal_pathlib) later on.
- Adds tests for Zarr datasets
- Renames `wkwResolutions` to `mags` in `datasource-properties.json` for Zarr datasets. cc @fm3

### Issues:
- fixes #622 

### Todos:
Make sure to delete unnecessary points or to check all before merging:
 - [x] Updated Changelog

normanrz commented on webknossos/webknossos/dataset/layer.py:
```diff
@@ -261,13 +265,17 @@ def get_best_mag(self) -> MagView:
         mag: Union[int, str, list, tuple, np.ndarray, Mag],
-        block_len: int = 32,
-        file_len: int = DEFAULT_WKW_FILE_LEN,
+        chunk_size: Optional[Union[Vec3IntLike, int]] = None,  # DEFAULT_CHUNK_SIZE,
```
Not sure if these functions should accept `Optional[Union[Vec3IntLike, int]]` or `Optional[Vec3IntLike]`

normanrz commented on webknossos/webknossos/dataset/view.py:
```diff
@@ -48,15 +56,27 @@ def __init__(
         self._mag = mag
 
+    @property
+    def info(self) -> StorageArrayInfo:
```
Not sure if this should be cached as well?

normanrz commented on webknossos/webknossos/dataset/storage.py:
```diff
@@ -0,0 +1,360 @@
+        return None
+
+    @property
+    def info(self) -> StorageArrayInfo:
```
Not sure if this should be cached? Same goes for https://github.com/scalableminds/webknossos-libs/pull/627/files#diff-3fe81ffa3877537cf9b035880c7881ae77f6cf7ec59fd1a59c68206d71eeaddcR254

philippotto reviewed (commented):
Cool stuff! Haven't finished reviewing, but wanted to let some comments here already..

normanrz commented on webknossos/webknossos/dataset/storage.py:
```diff
@@ -0,0 +1,366 @@
+
+    @classmethod
+    @abstractmethod
+    def try_open(_cls, path: Path) -> Optional["StorageArray"]:
```
Good point, I'll rework that with an exception. I think this was influenced by my rust experiences.

normanrz commented on webknossos/tests/test_dataset.py:
```diff
@@ -1956,22 +1991,23 @@ def test_add_layer_like(tmp_path: Path) -> None:
 def test_pickle_view(tmp_path: Path) -> None:
     ds = Dataset(tmp_path / "ds", scale=(1, 1, 1))
     mag1 = ds.add_layer("color", COLOR_CATEGORY).add_mag(1)
+    del mag1._array
```
`add_mag` opens the dataset to set the bbox. then `_cached_array` is not `None` anymore. Not sure why the assertion in the next line was there in the first place.

philippotto replied:
ah, I see. I'd rather remove the `del` + the first `assert`. I assume that the high-level thought of this test was to also check that `_cached_array` is not serialized somehow to pickle (and as a prerequisite it was also tested that the `_cached_array` was constructed lazily). Keeping all but the first `assert` should work too as far as I understand.

fm3 commented on webknossos/webknossos/dataset/properties.py:
```diff
@@ -185,6 +197,78 @@ class DatasetProperties:
+        obj: Union[LayerProperties, SegmentationLayerProperties]
+    ) -> Dict[str, Any]:
+        d = converter_fn(obj)
+        if d["dataFormat"] == "wkw":
```
I think unifying this would be worth a lot. We should use the same term across webknossos, voxelytics and webknossos-libs. I don’t think it is super important whether it is Mag, Magnification, Mip, or Resolution, but we should decide which we want.  Compare https://github.com/scalableminds/webknossos/issues/6053

normanrz commented on webknossos/webknossos/dataset/storage.py:
```diff
@@ -0,0 +1,366 @@
+
+@dataclass
+class StorageArrayInfo:
+    data_format: StorageArrayFormat
```
It is called `data_format` in the datasource-properties. Maybe it should be kept consistent with that? `format` is a built-in function that probably shouldn't be redefined (the linter whines about that).

philippotto commented on webknossos/webknossos/dataset/storage.py:
```diff
@@ -0,0 +1,366 @@
+    voxel_type: np.dtype
+    chunk_size: Vec3Int
+    chunks_per_shard: Vec3Int
+    compression_mode: bool = False
```
sure, if the format is only constructed ad-hoc and doesn't need a "migration" later, doing it later is also fine :+1:

philippotto reviewed (changes requested):
Great job abstracting wkw and zarr :+1: The big picture already looks very good to me.


I reviewed everything except for the (huge) test file. Wouldn't it make sense to generalize the tests so that most of them simply run twice (once for wkw and once for zarr). I think, most of the tests don't deal with the implementation specifics. If some do, they could stay separate. What do you think?

normanrz commented on webknossos/webknossos/dataset/dataset.py:
```diff
@@ -473,21 +483,32 @@ def add_layer_like(self, other_layer: Layer, layer_name: str) -> Layer:
         assert layer_name not in self.layers, f"Layer {layer_name} already exists!"
-        mag_headers = list((self.path / layer_name).glob("*/header.wkw"))
+
+        storage_info = None
+        for f in (self.path / layer_name).iterdir():
+            if f.is_dir():
+                try:
+                    array = StorageArray.open(f)
+                    storage_info = array.info
+                    break
+                except StorageArrayException:
+                    continue
```
Haha. This is what I had `StorageArray.try_open` for :-)
Extracting is a good idea, but it seems to specialized to me for to put in `StorageArray` class

normanrz replied:
I extracted a `_find_storage_array_info` function.

normanrz commented on webknossos/webknossos/dataset/layer.py:
```diff
@@ -338,24 +408,51 @@ def get_or_add_mag(
-        )
+        compression_mode = compress
+
+        if chunk_size is not None:
+            if isinstance(chunk_size, int):
+                chunk_size = Vec3Int.full(chunk_size)
+            else:
+                chunk_size = Vec3Int(chunk_size)
+        elif chunk_size is None and block_len is not None:
+            warnings.warn(
+                "[DEPRECATION] `block_len` is deprecated, please use `chunk_size` instead.",
+                DeprecationWarning,
+            )
+            chunk_size = Vec3Int.full(block_len)
+
+        if chunks_per_shard is not None:
+            if isinstance(chunks_per_shard, int):
+                chunks_per_shard = Vec3Int.full(chunks_per_shard)
+            else:
+                chunks_per_shard = Vec3Int(chunks_per_shard)
+        elif chunks_per_shard is None and file_len is not None:
+            warnings.warn(
+                "[DEPRECATION] `file_len` is deprecated, please use `chunks_per_shard` instead.",
+                DeprecationWarning,
+            )
+            chunks_per_shard = Vec3Int.full(file_len)
```
I think it is slightly different in the different cases, but I'll check what can be extracted.

normanrz replied:
This was used 3 times. Only once, is was used with default values instead of `None`. So, after all, this was easy to extract.

normanrz commented on webknossos/webknossos/dataset/storage.py:
```diff
@@ -0,0 +1,398 @@
+
+        new_shape_tuple = (
+            zarray.shape[0],
+            max(zarray.shape[1], new_shape.x),
+            max(zarray.shape[2], new_shape.y),
+            max(zarray.shape[3], new_shape.z),
```
Changed to `ensure_size`

normanrz commented:
> I reviewed everything except for the (huge) test file. Wouldn't it make sense to generalize the tests so that most of them simply run twice (once for wkw and once for zarr). I think, most of the tests don't deal with the implementation specifics. If some do, they could stay separate. What do you think?

I reworked `test_dataset.py` and dropped `test_zarr_dataset.py`, again.

normanrz commented:
I think the remaining points from your review are:
- `del mag1._array` in `test_dataset.py`
- `data_format` or `array_format` consistency
- `resolutions`, `magnifications`/`mags`, `multiscales`

philippotto commented on webknossos/tests/test_dataset.py:
```diff
@@ -255,10 +282,10 @@ def test_view_write() -> None:
 
 def test_view_write_out_of_bounds() -> None:
-    new_dataset_path = TESTOUTPUT_DIR / "wk_view_dataset_out_of_bounds"
+    new_dataset_path = TESTOUTPUT_DIR / "wkw_view_dataset_out_of_bounds"
```
wouldn't the test also make sense for zarr?

philippotto commented on webknossos/tests/test_dataset.py:
```diff
@@ -313,9 +343,9 @@ def test_mag_view_write_out_of_bounds_mag2() -> None:
 
 def test_update_new_bounding_box_offset() -> None:
-    delete_dir(TESTOUTPUT_DIR / "wk_dataset")
+    delete_dir(TESTOUTPUT_DIR / "wkw_dataset")
```
same here

philippotto commented on webknossos/tests/test_dataset.py:
```diff
@@ -292,10 +322,10 @@ def test_mag_view_write_out_of_bounds() -> None:
 
 def test_mag_view_write_out_of_bounds_mag2() -> None:
-    new_dataset_path = TESTOUTPUT_DIR / "simple_wk_dataset_out_of_bounds"
+    new_dataset_path = TESTOUTPUT_DIR / "simple_wkw_dataset_out_of_bounds"
```
same here

philippotto commented on webknossos/tests/test_dataset.py:
```diff
@@ -404,9 +444,9 @@ def test_read_padded_data() -> None:
 
 def test_num_channel_mismatch_assertion() -> None:
-    delete_dir(TESTOUTPUT_DIR / "wk_dataset")
+    delete_dir(TESTOUTPUT_DIR / "wkw_dataset")
```
same here

philippotto commented on webknossos/tests/test_dataset.py:
```diff
@@ -448,63 +497,84 @@ def test_get_or_add_layer() -> None:
 
 def test_get_or_add_mag() -> None:
-    delete_dir(TESTOUTPUT_DIR / "wk_dataset")
+    delete_dir(TESTOUTPUT_DIR / "wkw_dataset")
```
same here

philippotto reviewed (commented):
nice :+1: I think, more of the tests could be generalized to work on wkw and zarr, but I understand that it can be quite a hassle to convert this and the returns are probably diminishing. So, I don't have a hard opinion on that.

philippotto commented:
> * del mag1._array in test_dataset.py

I replied to the corresponding comment.

> * `data_format` or `array_format` consistency

Is this something you need an opinion on? I don't really have a tendency here :shrug: 

> * `resolutions`, `magnifications`/`mags`, `multiscales`

I'd vote for mags, as it's mainly used in newer code, concise and a bit clearer than the others (e.g., dataset resolution could also mean the physical size of a voxel (which we usually but not always refer to as scale)). I see, that it's not necessarily the most common term in the industry, but there will always be naming differences between different organizations/projects (see output ~ artifacts ~ debris). Being consistent within our own ecosystem is what matters most in my opinion.

normanrz commented on webknossos/webknossos/dataset/storage.py:
```diff
@@ -0,0 +1,408 @@
+
+@dataclass
+class StorageArrayInfo:
+    array_format: StorageArrayFormat
```
The problem is that `data_format` (or `dataFormat`) is currently used in the `datasource-properties.json`. This is the consistency conflict, I mentioned in the open questions.

jstriebel replied:
Ah, sorry, I didn't read the closed conversations before. This would speak for `data_format` then :shrug:

normanrz commented on webknossos/webknossos/dataset/storage.py:
```diff
@@ -0,0 +1,408 @@
+        return self.chunk_size * self.chunks_per_shard
+
+
+class StorageArray(ABC):
```
I think it `StorageArray` is not too far-fetched, because this layer of abstraction is supposed to be the closest layer to the actual persistance in this library. But, we can call it `Array` or `BaseArray`. Or even `DataArray` to fit the `data_format` attribute :trollface:

jstriebel reviewed (commented):
IMO this is good to go :tada:, just the respective Changelog entries are still needed. Leaving the final go to @philippotto :checkered_flag:

philippotto reviewed (approved):
Excellent :100: Good to go from my side, too.
